### DISCUSSION TITLE
What is the rational behind making v0.34 Type.Ref not allow passing in a schema and inferring the type?

### DISCUSSION THREAD

**Discussion by stuft2** [sinclairzx81/typebox#1071] [Q&A] [answered]
The reason I've used Type.Ref is to shorten the generated json schemas. In some cases it removes thousands of lines of repeated code.

Why add the breaking change to Type.Ref in v0.34? Why not allow backwards compatibility? Why not support type inference of reference types?

Wrapping Type.Ref with Type.Unsafe to get type inference is at least a workaround but I'd really like to understand what made this change necessary in the first place.

Thanks!

**Comment by sinclairzx81** [comment, marked as answer]
@stuft2 Hi,

> Why add the breaking change to Type.Ref in v0.34? 

The change was made in support of the new Module feature which provides enhanced (and Safe) inference support for Ref. Unfortunately the previous Ref  signature wasn't able to be repurposed in a way I felt comfortable maintaining moving forward (given conflicting designs). The updated signature on 0.34.0 is able to be made backwards compatible to 0.33.0, but the inverse wasn't possible. The call was made to drop the previous signature, but provide the fallback option via Unsafe.

> Why not allow backwards compatibility? 

You can replicate the old TRef using Unsafe (the following should work as before)

```typescript
import { Type, TSchema, Static, TUnsafe } from '@sinclair/typebox'

// Old Ref
function Ref<T extends TSchema>(schema: T): TUnsafe<Static<T>> {
  return Type.Unsafe<Static<T>>(Type.Ref(schema.$id!))
}

// ...

const T = Type.String({ $id: 'T' })

const R = Ref(T) // { $ref: 'T' }
```

> Why not support type inference of reference types?

Type inference is supported for Ref, but only within the context of a Module where it's possible to prove to the type system that the target type is correct.

```typescript
const Module = Type.Module({

  A: Type.Number(),             // Auto assigned { $id: 'A' } from property name
  
  R: Type.Ref('A'),             // { $ref: 'A' }
})

// ... 

type R = Static<typeof R>       // type R = number

const R = Module.Import('R')
```

Inference is also supported for mutually and self recursive types (something that wasn't possible before under the previous TRef design)

```typescript
const Module = Type.Module({
  Node: Type.Object({
    id: Type.String(),
    nodes: Type.Array(Type.Ref('Node'))
  })
})

// ... 

type Node = Static<typeof Node>       //  type Node = {
                                      //    id: string;
                                      //    nodes: ...[];
                                      // }

const Node = Module.Import('Node')
```

---

### Future

The updates to Module and Ref are part of a larger body of work to establish better alignment with the TS type system, specifically to enable order independent type definitions to reference (to be) defined types. The updates to Ref enable this (as they just accept strings to remote targets). By consequence of this, Ref is able to be used as a parse target for the new Syntax Type feature (which also support out of order type referencing)

[TypeScript Link Here](https://www.typescriptlang.org/play/?moduleResolution=99&module=199#code/JYWwDg9gTgLgBAbzgZRgQxsAxnAvnAMyghDgHIABAZ2ADssAbNYKAehgE8wBTAIwgAeZAFChIsRHAAKaKFW55CxUpRr0mLdlz6DWVDrXRDhw1qzgAhOFG4FuN+typwAgnF63oCt8GcATWzpuP2EsCFoqeAAxCAg4AF5pWXkACgADEAg-AFcGBRi4hBM4OE4eSwThEpKpbABrAB4XABpyYD8yAD44ADI4KuqZWGA0BgaAeRBgGCbWsnauzuLS7VcExAGS9oAuOEioOgBzTbhaLKddlwBtAF0B3BNcNIBKEzCI+CtEgoA6AElwNAYCkyBYyK9hGUFF8UOhMFgGlCIARLN1quiMWYVuUYUUMfiCYSidUseidnsYAdaIcANwnYkMxlwLH4Pp4pkc4mk6pnAJUAD8u3ZnJFmPMhPJ+yOdNFspJ4oJvIucB+qtuMrlsu5JVwtzgAB84NlaAECEE-BrNSKWUA)

```typescript
import { Static } from '@sinclair/typebox'
import { Parse } from '@sinclair/typebox/syntax'

// B references A before A is defined
const Foo = Parse(`module Foo {

  type B =
    Pick<A, 'id'> & 
    Partial<Omit<A, 'id'>>

  type A = {
    id: string
    nodes: A[]
  }

}`)

const B = Foo.Import('B')

type B = Static<typeof B>             // type B = {
                                      //     id: string;
                                      // } & {
                                      //     nodes?: {
                                      //         id: string;
                                      //         nodes: ...[];
                                      //     }[] | undefined;
                                      // }
````

The Module, Ref (Computed) and Syntax features will be expanded upon throughout 0.34.x. 

---

I apologize if this change has caused any inconvenience. I had been hesitant to implement it for several months but ultimately decided to proceed because I believe the new Module features address more problems than the breaking change to Ref. I hope the update hasn't caused too many problems

Happy to discuss more on this thread.
Cheers
S

**Reply by sinclairzx81** [reply]
@stuft2 FYI, Have just published a revert of the Ref(Schema) signature under a `@deprecated` comment. This is available on 0.34.7. There had been a few users impacted by the signature change, so in response to that, I'll keep the signature around for the 0.34.x revisions, and look at dropping in the next minor revision. Hopefully this helps smooth over the changes.

Cheers
S

**Reply by stuft2** [reply]
Due to the fact that module schemas can reference other schemas within the same module, do I still need to pass references into the `TypeCompiler.Compile()` function when trying to create a TypeCheck for a schema from within the module?

Example:
```ts
const Module = Type.Module({
  A: Type.Number(),             // Auto assigned { $id: 'A' } from property name
  R: Type.Ref('A'),             // { $ref: 'A' }
})

// Is this correct?
const RTypeCheck = TypeCompiler.Compile(Module.Import('R'))

// Or must I pass in the referenced schema?
const RTypeCheck = TypeCompiler.Compile(Module.Import('R'), [Module.Import('A')])
```

My guess is that it's unnecessary now, but I wanted to make sure since I didn't see anything in the documentation about Modules being used in the TypeCompiler

**Reply by stuft2** [reply]
Can modules be merged together? I have a use case where one package imports and uses another's schema module but it also needs to add other schemas to it to extend the original packages functionality.

**Reply by stuft2** [reply]
My current workaround is to expose the module schema properties from package A and then import and spread them into package B's schema module.

**Comment by liu-ronny** [comment]
@sinclairzx81 Is your recommendation to strictly use the new Module feature going forward when needing to use Ref? If so, can you offer some guidance on how to go about doing this?

I'm struggling with two problems.

### Schema reuse

I have some schemas (e.g., enums) that are reused by other schemas that may not be related to each other. I can't use Ref unless I do one of the following:

1. Redefine the referenced schemas within the relevant Modules 
2. Define one giant Module that holds all my schemas
3. Use Unsafe as a workaround

None of these options feel quite right. Or maybe I'm missing something obvious?

### Auto-generated OpenAPI documentation

I'm using TypeBox in a Fastify application, where I use plugins to generate OpenAPI documentation for my API routes. My understanding is that a Module is a schema with a `$defs` property, and using Import results in an object with all those `$defs` and a `$ref` to the specified subschema. How do I pair that with [`fastify.addSchema `](https://fastify.dev/docs/latest/Reference/Server/#addschema) to make sure my OpenAPI docs are correct?

**Comment by sinclairzx81** [comment]
@liu-ronny Hi,

>  Is your recommendation to strictly use the new Module feature going forward when needing to use Ref? If so, can you offer some guidance on how to go about doing this?

Ref is still intended to be used in isolation without Module (so you don't need to upgrade schematics to use Modules). However previous (and future) implementations using Ref will need to be mindful of a few subtle aspects with regards to the Ref API.

```typescript
const T = Type.Object({
  x: Type.Number(),
  y: Type.Number()
}, { $id: 'T' })

// ------------------------------------------
// Before 0.34.0
// ------------------------------------------

const R = Type.Ref(T)                                     // const R = { $ref: 'T' }

type R = Static<typeof R>                                 // type R = { x: number, y: number }

// ------------------------------------------
// After 0.34.0
// ------------------------------------------

const R = Type.Ref('T')                                   // const R = { $ref: 'T' }

type R = Static<typeof R>                                 // type R = unknown
```

CHANGES: The 0.34.0 version of Ref requires you to pass the $ref parameter via `string` (where-as before the $ref from read from the $id property of the schema). Additionally, because Ref now accepts a string argument, the inferred type for Ref must now be `unknown` as the Ref has no information about the target type.

## Ajv & Fastify

By updating Ref to take the string parameter, this is all that's required to make things work from a runtime / validation standpoint (using `fastify.addSchema(...)` etc). However type inference will be lost. See sections below.

## Type Inference

To retain inference, the recommended approach is to wrap the Ref in an Unsafe. The Unsafe type is used to override the default inference for any given TypeBox type. In the case of Ref, we need to use Unsafe to override the default inference of `unknown` to be that of the target type.

```typescript
// ------------------------------------------
// Before 0.34.0
// ------------------------------------------

const R = Type.Ref(T)                                     // const R = { $ref: 'T' }

type R = Static<typeof R>                                 // type R = { x: number, y: number }

// ------------------------------------------
// After 0.34.0
// ------------------------------------------

const R = Type.Unsafe<Static<typeof T>>(Type.Ref('T'))    // const R = { $ref: 'T' }

type R = Static<typeof R>                                 // type R = { x: number, y: number }
```

## UnsafeRef

As with other TypeBox types, you can construct a generic utility type that can provide a factory for the wrapped Unsafe Ref. The following reimplements pre-0.34.0 Ref via the utility type UnsafeRef.

```typescript
import { Type, TSchema, TUnsafe, Static } from '@sinclair/typebox'

// Replicates pre-0.34.0 Ref
function UnsafeRef<Type extends TSchema>(type: Type): TUnsafe<Static<Type>> {
  return Type.Unsafe<Static<Type>>(Type.Ref(type.$id!))
}

// ...

const T = Type.Object({
  x: Type.Number(),
  y: Type.Number(),
}, { $id: 'T' })

const R = UnsafeRef(T)                                    // const R = { $ref: 'T' }

type R = Static<typeof R>                                 // type R = { x: number, y: number }
```

## Why Ref changed

The main reason Ref changed was due to the following not being possible with the previous API

```typescript
const R = Type.Ref(T) // Error: Cannot use variable before initialization.

const T = Type.String({ $id: 'T' })
```
This is solved in 0.34.0
```typescript
const R = Type.Unsafe<Static<typeof T>>(Type.Ref('T')) // this is fine.

const T = Type.String({ $id: 'T' })
```

... so, the main reason for the change is to enable non-order dependent types to be created with reference types. The change here paves the way for upgrades to TypeBox's Recursive type implementation (due on the next release). Some of these upgrades are available via Module where deferred dereferencing via string parameter is facilitating both singular and mutual recursive schemas & inference. 

Unfortunately, the change did force Ref to infer as `unknown` (requiring the Unsafe wrap). This is largely due to Ref types being repurposed as a kind of "deferred" / "lazy" look up when used in the context of other structures (namely Modules). When Ref is used insolation, the Unsafe wrapping facilitates an "explicit" and "immediate" (non-lazy) dereference of the target structure.

---

Hope this helps. There was a fair amount of consideration that went into the Ref type update, but outside the signature change (passing string instead of schemas), they should largely work as before. The need for the extra explicit mechanisms (Unsafe) may be ironed out in future revisions (for example, making UnsafeRef a built in type (possibly named something else)), but for now, you will need wrap the Ref in Unsafe manually.

Happy to discuss more on this thread if you have any other questions
Cheers!
S

**Reply by liu-ronny** [reply]
Thanks for the detailed response.

Let's see if I've got this right.

Ref can still be used outside the context of a Module. However, since v0.34.0, it only takes a `string` argument. Outside of a Module, the result of Ref infers as type `unknown`.

In order to get a type-safe Ref outside of a Module, we can write a utility function that uses `TUnsafe` to mimic the behavior of Ref before v0.34.0. This allows an underlying `$ref` to be inferred as the input schema type.

The Module feature was introduced to allow for out of order references and self/recursive references. Modules also enable further alignment with TypeScript (e.g., we can parse an out of order TypeScript definition into a TypeBox Module). There's no need to put all our schemas in Modules.

Is this an accurate understanding of how everything works?

I'm still a bit confused about how Module works with `fastify.addSchema`. Let's say I introduce a Module into my code. Below is a contrived example.

```ts
import { Type, TypeBoxTypeProvider } from '@fastify/type-provider-typebox';
import { FastifyPluginAsync } from 'fastify';

const routes: FastifyPluginAsync = async (fastify, opts) => {
  const server = fastify.withTypeProvider<TypeBoxTypeProvider>();

  const Module = Type.Module({
    A: Type.Number(),
    R: Type.Ref('A'),
  });

  server.get(
    '/my-endpoint',
    {
      schema: {
        response: {
          '200': Module.Import('R'),
        },
      },
    },
    async function (request, reply) {
      return 1;
    },
  );
};
```

The response is type-safe and works as intended with `TypeBoxTypeProvider`. However, if I look at the `openapi.json` file generated by [`fastify-swagger`](https://github.com/fastify/fastify-swagger), I see the below.

```json
{
  "description": "Default Response",
  "content": {
    "application/json": {
      "schema": {
        "$defs": {
          "A": {
            "type": "number",
            "title": "A"
          },
          "R": {
            "$ref": "#/components/schemas/def-29",
            "title": "R"
          }
        },
        "$ref": "#/components/schemas/def-30"
      }
    }
  }
}
```

Without adding anything from the Module using `addSchema`, the `def-29` and `def-30` schemas don't exist. As a result, my docs show a null response schema for the `/my-endpoint` route. If I try to add the entire Module, `addSchema` complains that it doesn't have an `$id`, which I can't add because there's no options argument for Module. 

Do you have a recommendation for what to do here? Is this a plugin-specific problem with `fastify-swagger`?

**Reply by sinclairzx81** [reply]
@liu-ronny  Hiya

---

> Is this an accurate understanding of how everything works?

Yes, you understand everything correctly.

---

> The response is type-safe and works as intended with TypeBoxTypeProvider. However, if I look at the openapi.json file generated by [fastify-swagger](https://github.com/fastify/fastify-swagger), I see the below.

If using Modules, do not specify the `$id` as it will be auto generated internal to the Module (matching the internal types property name). When specifying the `$ref`, it must reference one of the properties names inside of the Module. 

```typescript
const Module = Type.Module({
  A: Type.String(),                          // the $id of 'A' will be auto generated based on the property name.
  B: Type.Ref('A')                           // B can reference 'A' property name.
})

const B = Module.Import('B')                 // const B = {
                                             //   $defs: {
                                             //     A: { type: "string", $id: "A" },
                                             //     B: { $ref: "A", $id: "B" }
                                             //   },
                                             //   $ref: "B"  <- Import Name
                                             // }
```
So, the following reference would be invalid in the context of Modules.
```typescript
"$ref": "#/components/schemas/def-29",
```

If you use Modules (which you shouldn't need to), you are opting-in to let TypeBox organize, structure, reference and de-reference schematics for you. The feature works by placing hard constraints on how $defs schematics can be structured (the same as other TB types), and from this it's able to perform auto dereferencing, deferred inferencing on your behalf.

This probably isn't going to be compatible with existing open Ref implementations unless they conform to the same structures mandated by Module.

---

> Do you have a recommendation for what to do here? Is this a plugin-specific problem with fastify-swagger?

Yes, just don't use Modules :) Just keep on with Ref. Just create the following function.

```typescript
// Or name this function however you like.
function LegacyRef<Type extends TSchema>(type: Type): TUnsafe<Static<Type>> {
  return Type.Unsafe<Static<Type>>(Type.Ref(type.$id!))
}
```
Then update all existing `Type.Ref(schema)` instances to be `LegacyRef(schema)`.

```typescript
const T = Type.Object({ ... }, { $id: '/components/schemas/def-29' })

const R = LegacyRef(T) // Replacing Type.Ref(T)
```
This will replicate the pre-0.34.0 Ref. Nothing else should need to change.

---

Keep in mind that TypeBox is NOT deprecating Ref. Open Ref implementations (as per your setup) are going to be supported forever. You should perceive the Module functionality as an opt-in extension feature that utilizes Ref. It is not the "be all and end all" way of handling references in TypeBox. The only thing that has changed is the signature for Ref, but if have a look at the `LegacyRef` type above, you can see there really isn't much of a change at all. 

Let me know this helps. If you have a codebase where the Ref types are still causing you problems. I am happy to take a look and help you refactor the Type.Ref inline with the 0.34.0 update.

Keep me posted
S

**Reply by stuft2** [reply]
> Error: Cannot use variable before initialization.

I just ran into this problem today, so I'm glad I have a path forward. And thanks again for keeping the current Ref implementation around for a bit while we transition. I'll be doing that today it seems 😄

**Reply by MoJo2600** [reply]
I am using Typebox for fastify as well and have a large ammount of schemas defined. I switched all my `Type.Ref` to `LegacyRef` as written above. But exactly since version `0.34.1` I'm getting a transpile error `Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory` - It works with `0.34.0`. I'm currently trying to figure out where exactly the error is, but I'm not able to provide an example repository which would reproduce the error unfortunately.

**Reply by sinclairzx81** [reply]
@MoJo2600 Hi,

Hmmm, would likely need to see some code in order to get bearings on this. 

I guess one thing you could try is doing a grep find/replace for all instances of `LegacyRef` and rename back to `Type.Ref`. Run the compiler and let me know if you run into any issues. I put the function back in for 0.34.x behind the deprecated flag to give integrators time to update.

Just try this and let me know if it helps
Cheers
S

**Reply by MoJo2600** [reply]
I don't want to rule out, that there is something strange with my code but:

I checked out the current version of my code wich is using version `0.32.35` of typebox. If I only upgrade the typebox package and do not change any code. I get the error `Argument of type 'TObject<...>' is not assignable to parameter of type 'string'` up until `0.34.6` and from `0.34.7` on I get the `heap out of memory` exception.

I will try to maybe recreate the issue in an repository, but I I understand it correctly the future proof way would be to use modules. I only have an handful of shared schemas so creating a module would not be that hard.

**Reply by sinclairzx81** [reply]
@MoJo2600 Hiya,

> I will try to maybe recreate the issue in an repository, but I I understand it correctly the future proof way would be to use modules. I only have an handful of shared schemas so creating a module would not be that hard.

The Ref pattern is still very much supported so would encourage you to continue with this pattern (via LegacyRef() or other appropriate name). There shouldn't be anything causing `heap out of memory` errors by swapping to LegacyRef(), I suspect there is something else going on.

As another possible option you can try is to remove `node_modules` and reinstall TypeBox fresh. If multiple packages are depending on varying versions of TypeBox, this "could" technically be cause issues. Ideally you want to make sure there is only one version of TypeBox installed throughout your project. Since you've upgraded, multiple installed versions is possible, it's worth checking this to rule out any problems.

Outside of this, I'm not too sure what else to suggest. A repro would be helpful if possible.

**Reply by MoJo2600** [reply]
Ok, I think I traced it down to the schema that causes the issue. It is very wierd and far beyond my typescript knowledge and I'm really sorry to bother you with this.

I do have a schema that looks like this:

```typescript
export const Scope = <T extends TSchema>(T: T) =>
    Type.Record(
        T,
        Type.Object({
            description: Type.String(),
        })
    );
```

Compilation fails with version > `0.34.6` but **only** if the `tsconfig.json` contains the the setting `compilerOptions.composite: true`. When I remove the setting, compilation works again. When the setting is set to true, and I comment this particular schema, it is working as well. (We were using the gts tsconfig as a baseline and this config has the setting set to true)

A repo with a reproducable error can be found here: https://github.com/MoJo2600/typebox-test

**Reply by sinclairzx81** [reply]
@MoJo2600 Hey, thanks for the repro code.

So, this does look like a TypeBox issue (so will investigate). The issue is unrelated to Ref, rather TypeBox introduced support for computed modules types on `0.34.6`, and it appears the `Key` parameter is causing TS to expand indefinitely when the `Key` is a generic argument. 

Workaround:

You should be able to work around this issue by reimplementing the type explicitly. As follows.

```typescript
const UserTokenScopeGroups = Type.Object({
    accounts: Type.Object({
        description: Type.String(),
        scopes: Type.Record(Type.Union([ // reimplement the type this way
            Type.Literal('accounts:read'),
            Type.Literal('accounts:write'),
        ]), Type.Object({
            description: Type.String(),
        }))
    }),
})
```
I will do some digging and see if there's something I can do to optimize for this case.

Ill keep you posted
S

**Reply by MoJo2600** [reply]
Yes, it dawned to me, that the issue could not be the refs. Thanks for the sample! will try this asap.

**Reply by sinclairzx81** [reply]
@MoJo2600 Hiya,

Have pushed a hotfix for this on 0.34.20. Have tested against you repro and it should resolve memory exhaustion issue for you (that was a wild bug that one). 

---

The issue itself appeared to be related to how TypeBox was handling TComputed types for Record. A Computed type is one that accepts a TSchema type and remaps it to another form. Historically, Computed types have not worked for Ref types (as remote schematics are required to perform some operation), however Modules provided an opportunity to get this working...

```typescript

const M = Type.Module({
  T: Type.Object({ value: Type.Number() }),
  R: Type.Record(Type.String(), Type.Partial(Type.Ref('T'))) 
  //                                            |
  //                                            |
  //       Attempt to Partial a Ref. To handle this case, TypeBox needs to transform 
  //       the Partial<T> into a Computed<'Partial', [Ref<'T'>]> which defers evaluation 
  //       of the Ref until Import. On Import TypeBox will dereference the target (T) and apply 
  //       the Partial operation to it. The result of the operation is assigned to Record Value.
})
```
The above was implemented on `0.34.6`, however given some pre-existing complexity around Record signature mapping, it appears the above logic caused TypeScript to explode when trying to ascribe a generic parameter to either the Key or Value of the Record. For now, I've pulled out the above logic and will investigate a more optimal solution later. Finding this was an excellent catch btw.

Give 0.34.20 a try, let me know if that helps
Cheers
S

**Reply by MoJo2600** [reply]
Thanks a lot! Not sure if I should be happy or sad that I found such a bug. I love your Typebox library, but a simple package upgrade escalated into a multi hour endeavor to find the issue 😅 But I'm glad that I could be of some help at least. I'll try the new version tomorrow.

**Reply by sinclairzx81** [reply]
@MoJo2600 Update: Have reimplemented the removed logic for handling Computed Record Types on `0.34.21`. Functionality is retained as much as possible, but have disabled Computed Record Keys. I will need revise Computed Record Keys in a minor semver as the changes to re-enable would be breaking). Have tested `0.34.21` against your repro example, everything looks good.

Cheers

**Reply by MoJo2600** [reply]
@sinclairzx81 yes, can confirm that it is working in my production code. Thank you very much!
